subsys/bluetooth/host/hci_core.c

```diff
@@ -1510,8 +1510,12 @@ void bt_hci_le_enh_conn_complete_sync(struct bt_hci_evt_le_enh_conn_complete_v2
 
 	bt_addr_le_copy(&conn->le.init_addr, &peer_addr);
 
-	/* There is no random addr to get, set responder addr to local identity addr. */
-	bt_addr_le_copy(&conn->le.resp_addr, &bt_dev.id_addr[conn->id]);
+	if (IS_ENABLED(CONFIG_BT_PRIVACY)) {
+		conn->le.resp_addr.type = BT_ADDR_LE_RANDOM;
+		bt_addr_copy(&conn->le.resp_addr.a, &evt->local_rpa);
+	} else {
```

Comment:
In `bt_hci_le_enh_conn_complete` we have 
```c
		if (IS_ENABLED(CONFIG_BT_PRIVACY)) {
			conn->le.init_addr.type = BT_ADDR_LE_RANDOM;
			if (bt_addr_cmp(&evt->local_rpa, BT_ADDR_ANY) != 0) {
				bt_addr_copy(&conn->le.init_addr.a,
					     &evt->local_rpa);
			} else {
				bt_addr_copy(&conn->le.init_addr.a,
					     &bt_dev.random_addr.a);
			}
		}
```

Does this need to take `if (bt_addr_cmp(&evt->local_rpa, BT_ADDR_ANY) != 0)` into account?

Reply:
I considered it, I am not sure it is necessary, The code you quote is only used if legacy advertising is used. If extended advertising is used there is some handling in the adv set terminated event which is almost the same. 

```c
                      if (IS_ENABLED(CONFIG_BT_PRIVACY) &&
			    !atomic_test_bit(adv->flags, BT_ADV_USE_IDENTITY)) {
				/* Set Responder address unless already set */
				conn->le.resp_addr.type = BT_ADDR_LE_RANDOM;
				if (bt_addr_cmp(&conn->le.resp_addr.a,
						BT_ADDR_ANY) == 0) {
					bt_addr_copy(&conn->le.resp_addr.a,
						     &adv->random_addr.a);
				}
			} else {
				bt_addr_le_copy(&conn->le.resp_addr,
					&bt_dev.id_addr[conn->id]);
			}
```
where `bt_addr_copy(&conn->le.resp_addr.a, &evt->local_rpa);` has already ran in `bt_hci_le_enh_conn_complete`.

I do not know if there is a way to have `BT_PRIVACY=y` and a non-resolvable or static address to be used.

Reply:
I guess it won't hurt to use `bt_dev.random_addr` as a fallback in case `evt->local_rpa` is all zeroes.

Reply:
I don't see any way for the address stored in the controller to not be an RPA if `CONFIG_BT_PRIVACY=y`.

Reply:
Okay, so no chance needed?

Reply:
ping @hermabe

Reply:
I have not been able to conclude whether there are other cases not covered or not. As far as I can tell, an RPA will be used if `BT_PRIVACY` is enabled and the id address otherwise. This PR will fix that case, but there may be other that still dont work.